metrics_prometheus/metrics.go

```diff
@@ -0,0 +1,193 @@
+		BadGateway:                  registry.NewCounter("bad_gateways", "number of bad gateway errors received from backends"),
+		EmptyContentLengthHeader:    registry.NewCounter("empty_content_length_header", "number of requests with the empty content length header"),
+		BackendInvalidID:            registry.NewCounter("backend_invalid_id", "number of bad backend id errors received from backends"),
+		BackendInvalidTLSCert:       registry.NewCounter("backend_invalid_tls_cert", "number of tls certificate errors received from backends"),
+		BackendTLSHandshakeFailed:   registry.NewCounter("backend_tls_handshake_failed", "number of backend handshake errors"),
+		BackendExhaustedConns:       registry.NewCounter("backend_exhausted_conns", "number of errors related to backend connection limit reached"),
+		WebsocketUpgrades:           registry.NewCounter("websocket_upgrades", "websocket upgrade to websocket"),
+		WebsocketFailures:           registry.NewCounter("websocket_failures", "websocket failure"),
+		Responses:                   registry.NewCounterVec("responses", "number of responses", []string{"status_group"}),
+		RouteServicesResponses:      registry.NewCounterVec("responses_route_services", "number of responses for route services", []string{"status_group"}),
+		RoutingResponseLatency:      registry.NewHistogramVec("latency", "routing response latency in ms", []string{"component"}, meterConfig.RoutingResponseLatencyHistogramBuckets),
+		perRequestMetricsReporting:  perRequestMetricsReporting,
+	}
+}
+
+func (metrics *Metrics) CaptureRouteStats(totalRoutes int, msSinceLastUpdate int64) {
+	metrics.TotalRoutes.Set(float64(totalRoutes))
+	metrics.TimeSinceLastRegistryUpdate.Set(float64(msSinceLastUpdate))
+}
+
+func (metrics *Metrics) CaptureRegistryMessage(msg metrics.ComponentTagged, action string) {
+	metrics.RouteRegistration.Add(1, []string{msg.Component(), action})
+}
+
+func (metrics *Metrics) CaptureUnregistryMessage(msg metrics.ComponentTagged) {
+	metrics.RouteUnregistration.Add(1, []string{msg.Component()})
+}
+
+func (metrics *Metrics) CaptureRoutesPruned(routesPruned uint64) {
+	metrics.RoutesPruned.Add(float64(routesPruned))
+}
+
+func (metrics *Metrics) CaptureTotalRoutes(totalRoutes int) {
+	metrics.TotalRoutes.Set(float64(totalRoutes))
+}
+
+func (metrics *Metrics) CaptureTimeSinceLastRegistryUpdate(msSinceLastUpdate int64) {
+	metrics.TimeSinceLastRegistryUpdate.Set(float64(msSinceLastUpdate))
+}
+
+func (metrics *Metrics) CaptureLookupTime(t time.Duration) {
+	if metrics.perRequestMetricsReporting {
+		metrics.RouteLookupTime.Observe(float64(t.Nanoseconds()))
+	}
+}
+
+func (metrics *Metrics) CaptureRouteRegistrationLatency(t time.Duration) {
+	metrics.RouteRegistrationLatency.Observe(float64(t) / float64(time.Millisecond))
+}
+
+// UnmuzzleRouteRegistrationLatency should set a flag which suppresses metric data.
+// Deprecated: that makes sense for Envelope V1 where we send it to collector any time we got new value
```

Comment:
Please review the wording. As far as I understood, we don't deprecated Envelope V1 metrics overall for Gorouter. While some setups might not want to use them, they remain supported.